Setup web3-profile package

PaulLeCam opened the pull request:
Here's the setup I came up with so far, to be improved over time but that should cover our current use-cases:

- `schemas` folder contains the JSON schema sources, and using the `types` command the matching TS definitions are generated in the `types` folder.
- `src` folder contains the TS source, exports the `model` object that contains the signed model, as well as the `ModelTypes` definition.
- `dist` folder contains the compiled JS and type definitions.

michaelsena commented:
@PaulLeCam how do definitions fit into this? I just see you mention schemas?

PaulLeCam commented:
> @PaulLeCam how do definitions fit into this? I just see you mention schemas?

Definitions need to be created manually in the `ModelManager` provided by the devtools, eventually I think that could be made declarative like having a JSON or YAML file for example that defines the relations between the schemas and dependencies, so the process of creating the schemas and definitions could happen in CI, but that would be a longer-term goal.

The `basicProfile` definition commit is set [here](https://github.com/ceramicstudio/datamodels/pull/1/files#diff-f14298db6bade1d2a82cda0d74c7aa9c206f5bec16781fc9aa79370cc7b613fbR13) though, so it can be used at runtime already once the model is published.

michaelsena commented:
Okay another question. How do these data models relate to CIPs?

On Thu, Jul 15, 2021 at 8:23 AM Paul Le Cam ***@***.***>
wrote:

> @PaulLeCam <https://github.com/PaulLeCam> how do definitions fit into
> this? I just see you mention schemas?
>
> Definitions need to be created manually in the ModelManager provided by
> the devtools, eventually I think that could be made declarative like having
> a JSON or YAML file for example that defines the relations between the
> schemas and dependencies, so the process of creating the schemas and
> definitions could happen in CI, but that would be a longer-term goal.
>
> The basicProfile definition commit is set here
> <https://github.com/ceramicstudio/datamodels/pull/1/files#diff-f14298db6bade1d2a82cda0d74c7aa9c206f5bec16781fc9aa79370cc7b613fbR13>
> though, so it can be used at runtime already once the model is published.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceramicstudio/datamodels/pull/1#issuecomment-880650912>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHZQESYZ3OYCL45GUBWLITTX3HNRANCNFSM5ANKZNGA>
> .
>

oed commented on packages/web3-profile/package.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "@datamodels/web3-profile",
```
Wouldn't it make more sense to call this `basic-profile`, as that would create a 1-1 mapping to the CIP?

michaelsena replied:
Should we consider adding the format to the namespacing (e.g. JSON), so when Ceramic potentially supports other schema types that it's clearly in the name? Should we consider something like `json-profile-basic` or `profile-basic-json`

michaelsena replied:
Also, as implied above, should we put the descriptor after the object type (profile-basic) so that multiple profiles will show up next to each other in the package list? It's easy to imagine something like:
- `profile-basic-json`
- `profile-basic-cson`
- `profile-dao-json`
- `profile-gitcoin-json`
- `profile-user-json`
- ...

PaulLeCam replied:
I don't think we need to include the format, all we support is JSON schemas anyways so it's the entire stack that would need to be updated if Ceramic were to support additional ways to create schemas.
Regarding namespacing, personally I would find it more useful as semantic relations, ex `schema.org-action`, `schema.org-creative-work`, `w3c-activitypub` so it's easier to figure out the datamodels that are more likely to be working well together, otherwise I fear we might end-up with 100s of `profile-...` that are hard to differentiate.

oed replied:
Whats *CSON* 😹

oed replied:
Naming is hard here. I think what Paul suggests with domain names could make sense. It's annoying because the schemas and definitions themselves have unique identifiers (streamids).

michaelsena replied:
What other semantic relations/descriptors could we imagine? I guess a lot would default to using the web3 namespace?

PaulLeCam replied:
Yes the downside of `web3` is that it's such a large scope, that's why I thought the company/organization name could be better suited.
Eventually, I think we could have both wide-ranging namespaces (ex `w3c-activitypub`) and more focused ones (for projects that want to share their data models but still "own" them), so our convention should be more about avoiding misleading names.

Regarding our own datamodels, maybe we can scope them to the projects we're primarily using them for? So here maybe `self.id-profile` or just `self-profile`?

oed commented:
Looks good other than my question!

PaulLeCam commented:
> Okay another question. How do these data models relate to CIPs?

At the moment all the definitions and schemas we have come from CIPs, are you asking if this should be a requirement? If so it's really up to how open vs curated we would want this repository to be.

@oed depends if we want to make these CIP-centric or namespaced (ex `web3-profile`, `web3-crypto-accounts`, etc.)?
But overall I think we should set some naming rules or at least conventions to avoid name-squatting or other misleading practices, and using namespace prefixes could help there.

PaulLeCam commented:
@oed @michaelsena I changed the package name to `self.id-profile` and added `self.id-crypto-accounts`, `self.id-social-accounts` and `3id-keychain`, though we can rename these as you see fit.
Maybe `self.id-crypto-accounts` could be `3id-account-links` instead as they're used in the 3ID provider IIRC?

oed commented:
Using the `self.id` namespace makes sense to me. Using `3id-account-links` doesn't really make sense as any DID could use the crypto account definition, also not used in the 3id-provider anymore.

michaelsena commented:
I'm wondering how much it matters or is relevant at this point to work to try to namespace according to application/creator, when we're actively trying to move away from that model toward one that is data-centric and agnostic to application. If I was going to throw away that requirement (namespacing by creator), then I would imagine something like this being most intuitive:

`web3-profile`
`crypto-accounts`
`also-known-as`
`auth-keychain`

PaulLeCam commented:
I'm concerned the lack of namespacing would create a mess, misleading/name-squatting packages or simply overall limit the discovery. To me `web3-profile`, `crypto-accounts` and `also-known-as` don't have any obvious meaning without the context of being used by self.id or another service.
Eventually, it's likely we'll have different profiles for different services that need more specific schemas than the basic profile. What happens when we start having `profile`, `web2-profile`, `user-profile`...?
